src/annotator/anchoring/normalize.js

```diff
@@ -0,0 +1,224 @@
+    // Works around the fact that browsers don't generate
+    // ranges/selections in a consistent manner. Some (Safari) will create
+    // ranges that have (say) a textNode startContainer and elementNode
+    // endContainer. Others (Firefox) seem to only ever generate
+    // textNode/textNode or elementNode/elementNode pairs.
+    if (range.tainted) {
+      throw new Error('You may only call normalize() once on a BrowserRange!');
+    } else {
+      range.tainted = true;
+    }
+    const rangeTemp = {};
+    // Look at the start
+    if (range.startContainer.nodeType === Node.ELEMENT_NODE) {
+      // We are dealing with element nodes
+      if (range.startOffset < range.startContainer.childNodes.length) {
+        rangeTemp.start = getFirstTextNodeNotBefore(
+          range.startContainer.childNodes[range.startOffset]
+        );
+      } else {
+        rangeTemp.start = getFirstTextNodeNotBefore(range.startContainer);
+      }
+      rangeTemp.startOffset = 0;
+    } else {
+      // We are dealing with simple text nodes
+      rangeTemp.start = range.startContainer;
+      rangeTemp.startOffset = range.startOffset;
+    }
+
+    // Look at the end
+    if (range.endContainer.nodeType === Node.ELEMENT_NODE) {
+      // Get specified node.
+      let node = range.endContainer.childNodes[range.endOffset];
+      // Does that node exist?
+      if (node) {
+        // Look for a text node either at the immediate beginning of node
+        let n = node;
+        while (n && n.nodeType !== Node.TEXT_NODE) {
+          n = n.firstChild;
+        }
+        // Did we find a text node at the start of this element?
+        if (n) {
+          rangeTemp.end = n;
+          rangeTemp.endOffset = 0;
+        }
+      }
+
+      if (!rangeTemp.end) {
+        // We need to find a text node in the previous sibling of the node at the
+        // given offset, if one exists, or in the previous sibling of its container.
+        if (range.endOffset) {
+          node = range.endContainer.childNodes[range.endOffset - 1];
+        } else {
+          node = range.endContainer.previousSibling;
+        }
+        rangeTemp.end = getLastTextNodeUpTo(node);
+        rangeTemp.endOffset = rangeTemp.end.nodeValue.length;
+      }
+    } else {
+      // We are dealing with simple text nodes
+      rangeTemp.end = range.endContainer;
+      rangeTemp.endOffset = range.endOffset;
+    }
+
+    // We have collected the initial data.
+    // Now let's start to slice & dice the text elements!
+    const normalRange = {};
+    if (rangeTemp.startOffset > 0) {
+      // Do we really have to cut?
+      if (
+        !rangeTemp.start.nextSibling ||
+        rangeTemp.start.nodeValue.length > rangeTemp.startOffset
+      ) {
+        // Yes. Cut.
+        normalRange.start = rangeTemp.start.splitText(rangeTemp.startOffset);
+      } else {
+        // Avoid splitting off zero-length pieces.
+        normalRange.start = getFirstTextNodeNotBefore(
```

Comment:
This was not a straight conversion, the original coffeescript did not call getFirstTextNodeNotBefore(), it just set normalRange.start to rangeTemp.start.nextSibling. but nextSibling may not always be a text node. I discovered this issue while building out unit tests. I'm not 100% I'm right about this, but it could be a long standing bug, or just my misunderstanding.